lib/IRGen/GenKeyPath.cpp

```diff
@@ -1105,6 +1106,21 @@ emitKeyPathComponent(IRGenModule &IGM,
   case KeyPathPatternComponent::Kind::OptionalWrap:
     fields.addInt32(KeyPathComponentHeader::forOptionalWrap().getData());
     break;
+  case KeyPathPatternComponent::Kind::TupleElement:
+    assert(baseTy->is<TupleType>() && "not a tuple");
+          
+    SILType loweredTy = IGM.getSILTypes().getLoweredType(baseTy);
+
+    if (auto offset = getFixedTupleElementOffset(IGM, loweredTy, component.getTupleIndex())) {
+      auto header = KeyPathComponentHeader
+                      ::forStructComponentWithInlineOffset(/*isLet*/ false,
+                                                           offset->getValue());
+        
+      fields.addInt32(header.getData());
+      break;
+    }
```

Comment:
Given that Swift prefers small incremental changes for its evolution process, do we want include this feature in the PR or we prefer to implement this later?

Reply:
I think it would be good to implement the entire feature before accepting the change. If you want to formally propose the language change, though, I personally don't see a problem starting the proposal now even if the implementation isn't 100% complete. We know it's implementable.

Reply:
While waiting for xedin's reply to the other comment, I'm tackling this. I'm looking at the code using `forStructComponentWithUnresolvedFieldOffset` for `struct`'s and `class`'es in the `StoredProperty` case, but the code that creates / gets the metadata layout seems a little bit too magical to me to understand. I could found and understand what happens for nominal types, but the code in _GenTuple_ is not helping me getting a reference to a tuple metadata object. Any guidance?

Reply:
Your best bet reference for the tuple type metadata layout would be to look at the Metadata.h header:

https://github.com/apple/swift/blob/master/include/swift/ABI/Metadata.h#L1451

It looks like, after the three words for metadata kind, `NumElements`, and `Labels`, there's a trailing array of type/offset pairs for each of the elements. That would mean that the storage offset for element `n` is at byte offset `(4 + 2*n) * sizeof(void*)` within the metadata for the tuple type.

Reply:
@jckarter I've update the PR! I've found it useful to do a little refactoring and create a copule new classes, what do you think?